src/PointwiseFunctions/Hydro/EquationsOfState/Spectral.cpp

```diff
@@ -0,0 +1,317 @@
+  if constexpr (std::is_same_v<DataType, double>) {
+    return Scalar<double>{
+        specific_internal_energy_from_density(get(rest_mass_density))};
+  } else if constexpr (std::is_same_v<DataType, DataVector>) {
+    auto result = make_with_value<Scalar<DataVector>>(rest_mass_density, 0.0);
+    for (size_t i = 0; i < get(result).size(); ++i) {
+      get(result)[i] =
+          specific_internal_energy_from_density(get(rest_mass_density)[i]);
+    }
+    return result;
+  }
+}
+
+template <class DataType>
+Scalar<DataType> Spectral::chi_from_density_impl(
+    const Scalar<DataType>& rest_mass_density) const {
+  if constexpr (std::is_same_v<DataType, double>) {
+    return Scalar<double>{chi_from_density(get(rest_mass_density))};
+  } else if constexpr (std::is_same_v<DataType, DataVector>) {
+    auto result = make_with_value<Scalar<DataVector>>(rest_mass_density, 0.0);
+    for (size_t i = 0; i < get(result).size(); ++i) {
+      get(result)[i] = chi_from_density(get(rest_mass_density)[i]);
+    }
+    return result;
+  }
+}
+
+template <class DataType>
+Scalar<DataType> Spectral::kappa_times_p_over_rho_squared_from_density_impl(
+    const Scalar<DataType>& rest_mass_density) const {
+  return make_with_value<Scalar<DataType>>(get(rest_mass_density), 0.0);
+}
+
+// this evaluates the power series
+// Gamma(x) = Sum_{n=0}^N gamma_n x^n
+double Spectral::gamma(const double x) const {
+  double result = gamma_coefficients_.back();
+  for (size_t i = 1; i < gamma_coefficients_.size(); ++i) {
+    const size_t n = gamma_coefficients_.size() - i - 1;
+    result *= x;
+    result += gamma_coefficients_[n];
+  }
+  return result;
+}
+
+double Spectral::chi_from_density(const double rest_mass_density) const {
+  const double x = log(rest_mass_density / reference_density_);
+  const double P = pressure_from_log_density(x);
+  const double chi =
+      (x <= 0.0 ? integral_coefficients_[0] * P / rest_mass_density
+                : (x < x_max_ ? gamma(x) * P / rest_mass_density
+                              : gamma(x_max_) * P / rest_mass_density));
+  return chi;
+}
+
+double Spectral::specific_internal_energy_from_density(
+    const double rest_mass_density) const {
+  const double x = log(rest_mass_density / reference_density_);
+  if (x <= 0.) {
+    return reference_pressure_ / reference_density_ /
+           (gamma_coefficients_[0] - 1.0) *
+           exp((gamma_coefficients_[0] - 1.0) * x);
+  }
+  const size_t n_points_epsilon = table_of_specific_energies_.size();
+  const double delta_x = x_max_ / (n_points_epsilon - 1.0);
+  double specific_energy = 0.0;
+  if (x >= x_max_) {
+    // Analytic integral at constant gamma_coefficient above x_max_
+    specific_energy = table_of_specific_energies_[n_points_epsilon - 1];
+    specific_energy += pressure_from_log_density(x_max_) / reference_density_ *
+                       exp(-x_max_) *
+                       (exp((gamma_of_x_max_ - 1.0) * (x - x_max_)) - 1.0) /
```

Comment:
If this factor can be small, use `std::expm1`.

Reply:
From the documentation, it seems that std::expm1 is good to use if (-x_max) is close to 0 and we want (e^x-1). Is there a different function that is useful for large negative exponents?

Reply:
If the exponent is never close to zero what you have is fine.  (`expm1` would also be fine.)

Reply:
x_max should be of order of "a few" (maybe 10ish), so we should be fine.

Reply:
Maybe I'm missing something, but it seems like x-x_max can be small (line 211)?